docs/installation/upgrading-from-1.1.md

```diff
@@ -0,0 +1,49 @@
+# Upgrading DjNRO from 1.1 to 1.2 or later
+
+DjNRO 1.1 was running with django 1.8. Version 1.2 was updated for django 1.11
+
+The main change in the Django 1.11 upgrade is that the `python-social-auth` package is replaced with `social-auth-core`.
+
+This has effects on the database (see section on migrations below), and also affects `local_settings.py` (replace `social` namespace with `social_core`)
+
+Additionally, the `TEMPLATES_DEBUG` setting has been removed and template debugging is controlled by `DEBUG` - which was the default for `TEMPLATE_DEBUG` in `local_settings.py` anyway.
+
+## Backup data
+
+Dump data (always keep a backup):
+
+	./manage.py dumpdata --indent=4 --natural --exclude=contenttypes --exclude=auth.Permission > /tmp/eduroam_1.1.json
+
+And, possibly, make also a full database dump from your database, and also make
+an archive of your installation tree - to be able to revert if something goes
+wrong.
+
+
+## Install DjNRO
+
+Install DjNRO >= v1.2 by following the installation instructions - configuring it to use the same database.
+
+
+## Migrate
+
+### Social Auth Core
+
+The `python-social-auth` package has been renamed to `social-auth-core` (and has had sigificant changes done to it afterwards).
+
+The `social-auth-core` package has kept the same sequence of database migration steps, but because of the rename, Django would get confused and would try applying the migrations from start ... which would fail, because the tables already exist.
+
+Apply the following workaround: change the package name in the migrations history to make Django Migrations think these have been applied under the new package name.   (Because of the hierarchy used in the old package versions, the migrations were recorded as belonging to package `default`).  Run `./manage.py shell` and paste in the following code:
+
+    from django.db.migrations.recorder import MigrationRecorder
+    for m in MigrationRecorder.Migration.objects.filter(app='default'):
+        if not MigrationRecorder.Migration.objects.filter(app='social_auth', name=m.name):
+            print("Recording migration default:%s as social_auth:%s" % (m.name, m.name))
+            m_social=MigrationRecorder.Migration(app='social_auth', name=m.name)
```

Comment:
I'm not sure I understand why we would need to manually update applied migrations with `MigrationRecorder`, considering that PSA added the `replaces` attribute in https://github.com/omab/python-social-auth/commit/361f42235fe51f422af029f325d041407a16036c for migrations affected by the django app rename (`default` to `social_auth`, and this was carried over to `social_django`). Did you find this did not work for some reason?

Reply:
Hi @zmousm , I also found the "replaces" annotation ... but wasn't working for me, and Django was trying to apply the new chain of migrations from scratch ... and was failing on that (that part is to be expected).

I'm not exactly sure why the "replaces"  annotation wasn't applying ... maybe because the original application did not load anymore?

I could do some experiments to confirm this - but I added it into the documentation because otherwise the migrations were failing...

Cheers,
Vlad

Reply:
I had a closer look at the issue, spending many hours navigating the infinite complexity of Django in order to understand the cause.

In my experience the problem is such a backtrace while running migrations:

```
  Applying social_django.0004_auto_20160423_0400...Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/db/migrations/migration.py", line 119, in apply
    operation.state_forwards(self.app_label, project_state)
  File "/srv/venv/djnro_django111/local/lib/python2.7/site-packages/django/db/migrations/operations/fields.py", line 201, in state_forwards
    state.models[app_label, self.model_name_lower].fields
KeyError: ('social_django', u'usersocialauth')
```

Searching about that will give you https://github.com/python-social-auth/social-app-django/issues/28, where the suggested workaround is to first upgrade to PSA 0.2.21, as an intermediate step. In our case you did this in 9ff2496, where the Django 1.11 upgrade is already under way. I think it should have been committed (and tagged) before this PR. We could then guide users to first checkout that tag, `pip install -r ...` and run migrations, per the suggested workaround, before going further. Nevertheless we can still instruct them how to do this manually.

Your suggested workaround involving `MigrationRecorder` also works, of course, and I admit it is more simple. In principle however I think it is best to avoid juggling with Django internals for user-oriented guidelines.

I have updated the upgrading guide, providing an explanation for the issue as well as both workarounds, and pushed this to #58.

Reply:
The actual reason for this backtrace is as follows:

When jumping over PSA 0.2.21, the migrations for `social_auth` are not recorded in the database. `social_django` migrations are declared as _replacements_ for respective migrations in both `default` and `social_auth` apps. Django however requires that _**all** replacement targets are marked as applied for the replacing migration to be marked as applied_. See https://github.com/django/django/blob/331d76528154407927f88759c9e520494e29b914/django/db/migrations/loader.py#L229. Failing this, `social_django` is excluded from the project state, and this leads to the `KeyError` in question. The connection between the two is quite loose, admittedly, but I think one could even argue this constitutes a bug (or two) in Django.

Reply:
Hi @zmousm , thanks a lot for this code deep dive.
Yes, you are right - I initially started with updating to last PSA (0.2.21).
In the end, I decided for the workaround I put into the documentation as easier to follow (though I'm neither a fan of touching Django internals either).  Happy with your other workaround as well.
Thanks again for all the hard work put into this!
Cheers,
Vlad